File: src/main/java/com/leet/medium/TrimABinarySearchTree.exs
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec trim_bst(root :: TreeNode.t() | nil, low :: integer, high :: integer) ::
          TreeNode.t() | nil
  def trim_bst(root, low, high) do
    if root == nil do
      nil
    else
      cond do
        root.val < low ->
          trim_bst(root.right, low, high)

        root.val > high ->
          trim_bst(root.left, low, high)

        true ->
          %TreeNode{
            val: root.val,
            left: trim_bst(root.left, low, high),
            right: trim_bst(root.right, low, high)
          }
      end
    end
  end
end

File: src/main/java/com/leet/medium/TrimABinarySearchTree.rs
use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None,
    }
  }
}

impl Solution {
  pub fn trim_bst(
    root: Option<Rc<RefCell<TreeNode>>>,
    low: i32,
    high: i32,
  ) -> Option<Rc<RefCell<TreeNode>>> {
    return match (root) {
      None => root,
      Some(node_ref) => {
        let mut node = node_ref.borrow();
        if node.val < low {
          return Solution::trim_bst(node.right.clone(), low, high);
        }
        if node.val > high {
          return Solution::trim_bst(node.left.clone(), low, high);
        }
        let ans = TreeNode {
          val: node.val,
          left: Solution::trim_bst(node.left.clone(), low, high),
          right: Solution::trim_bst(node.right.clone(), low, high),
        };
        return Some(Rc::new(RefCell::new(ans)));
      }
    };
  }
}
